abci/multiplexer.go

```diff
@@ -46,42 +82,132 @@ func NewVersions(v ...Version) (Versions, error) {
 }
 
 // NewMultiplexer creates a new ABCI wrapper for multiplexing
-func NewMultiplexer(
-	logger log.Logger,
-	v *viper.Viper,
-	latestApp servertypes.ABCI,
-	versions Versions,
-	chainID string,
-	applicationVersion uint64,
-) (proxy.ClientCreator, func() error, error) {
-	var noOpCleanUp = func() error {
-		return nil
-	}
-
+func NewMultiplexer(svrCtx *server.Context, svrCfg serverconfig.Config, clientCtx client.Context, appCreator servertypes.AppCreator, versions Versions, chainID string, applicationVersion uint64) (*Multiplexer, error) {
 	if err := versions.Validate(); err != nil {
-		return nil, noOpCleanUp, fmt.Errorf("invalid versions: %w", err)
+		return nil, fmt.Errorf("invalid versions: %w", err)
 	}
 
-	wrapper := &Multiplexer{
-		logger:         logger,
-		latestApp:      latestApp,
+	mp := &Multiplexer{
+		svrCtx:         svrCtx,
+		svrCfg:         svrCfg,
+		clientContext:  clientCtx,
+		appCreator:     appCreator,
+		logger:         svrCtx.Logger.With("multiplexer"),
+		latestApp:      nil, // app will be initialized if required by the multiplexer.
 		versions:       versions,
 		chainID:        chainID,
 		lastAppVersion: applicationVersion,
+		cleanupFns:     make([]func() error, 0),
+	}
+
+	return mp, nil
+}
+
+// isGrpcOnly checks if the GRPC-only mode is enabled using the configuration flag.
+func (m *Multiplexer) isGrpcOnly() bool {
+	return m.svrCtx.Viper.GetBool(flagGRPCOnly)
+}
+
+// registerCleanupFn enables the registration of additional cleanup functions that get called during Cleanup
+func (m *Multiplexer) registerCleanupFn(cleanUpFn func() error) {
+	m.cleanupFns = append(m.cleanupFns, cleanUpFn)
+}
+
+func (m *Multiplexer) Start() error {
+	m.g, m.ctx = getCtx(m.svrCtx, true)
```

Comment:
the reason I have added the waitgroup and ctx to the struct is because when we dynamically switch from an embedded app to a native app in the `getApp` function, we need to start the grpc/api services there. We need access to the same wait group and context I believe to keep the main foreground process behaving as normal. I need to test this more.